### PR TITLE
Node state machine

### LINKED ISSUE #2627: Document node state machine

### PR DESCRIPTION
**Pull request by achamayou**
Close #2627

### CODE REVIEW

File: doc/audit/node_state.rst [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+        Uninitialized-- config -->Initialized;
+        Initialized-- start -->PartOfNetwork;
+        Initialized-- join from snapshot -->VerifyingSnapshot;
+        VerifyingSnapshot-->Pending;
+        Initialized-- join -->Pending;
+        Pending-- poll status -->Pending;
+        Pending-- trusted -->PartOfNetwork;
+        Initialized-- recovery -->ReadingPublicLedger;
+        ReadingPublicLedger-->PartOfPublicNetwork;
+        PartOfPublicNetwork-- shares reassembly -->ReadingPrivateLedger;
```
Comment (jumaffre) [inline doc/audit/node_state.rst:16]:
```suggestion
        PartOfPublicNetwork-- member recovery shares reassembly -->ReadingPrivateLedger;
```

If this doesn't break the formatting of the graph.

**Review by jumaffre — APPROVED** [review]
Could we document the service state as well (which gets recorded in the service table). Something like:

```
.. mermaid::

    graph TB;
        Opening-- config -->Open;
        Opening-- transition_service_to_open-->WaitingForRecoveryShares;
        WaitingForRecoveryShares -- member recovery shares reassembly--> Open;
```

**Review by eddyashton — APPROVED** [review]

File: doc/audit/node_state.rst [hunk 2]
```diff
@@ -0,0 +1,17 @@
+Node State
+==========
+
+.. mermaid::
+
+    graph TB;
```
Comment (jumaffre) [inline doc/audit/node_state.rst:6]: Looking at this again, could we split the graph for each node (start, join, recover)? Also, as this is not recorded in the ledger, should this be moved to the operations section?

Comment (achamayou) [reply doc/audit/node_state.rst:6]: Very good points!

**Comment by achamayou** [comment]
@jumaffre yes!

image

File: doc/audit/builtin_maps.rst [hunk 3]
```diff
@@ -159,10 +159,20 @@ Service identity and status.
...
+
 .. doxygenstruct:: ccf::ServiceInfo
    :project: CCF
    :members:
 
+.. mermaid::
+
+    graph TB;
+        Opening-- config -->Open;
+        Opening-- transition_service_to_open-->WaitingForRecoveryShares;
```
Comment (jumaffre) [inline doc/audit/builtin_maps.rst:173]:
```suggestion
        Opening-- transition_service_to_open -->Open;
        Opening-- transition_service_to_open (recovery)-->WaitingForRecoveryShares;
```

My bad.

File: doc/operations/recovery.rst [hunk 4]
```diff
@@ -30,8 +30,27 @@ To initiate the first phase of the recovery procedure, one or several nodes shou
     recover
     --network-cert-file /path/to/network_certificate
 
+.. mermaid::
+
+    graph LR;
+        Uninitialized-- config -->Initialized;
```
Comment (jumaffre) [inline doc/operations/recovery.rst:36]: I don't think `Uninitialized` is actually observable as it is called from the original ECALL (in fact looking at the code, the entire `node->initialize` call could be merged in the `NodeState` constructor).

Comment (achamayou) [reply doc/operations/recovery.rst:36]: True, but the state exists internally, so I'm inclined to keep it for now until we've closed #857.

**Pull request merged**